Add `unmatch_mapping_type`, and support array of types

axw opened the pull request:
Add an `unmatch_mapping_type` condition to dynamic templates (supporting one or more types), and add support for specifying a list of types to `match_mapping_type`.

Closes https://github.com/elastic/elasticsearch/issues/102795
Closes https://github.com/elastic/elasticsearch/issues/102807

Commit:
Add unmatch_mapping_type; support array of types

Add an unmatch_mapping_type condition to dynamic templates,
and add support for specifying a list of types to
match_mapping_type.

axw commented:
@felixbarny @javanna I took a stab at addressing the linked issues. Could you please tell me if this looks directionally sensible, and if you can spot any major issues? If it looks alright I'll proceed to add docs and probably expand the tests a bit.

The main thing I'm unsure about is backwards compatibility with respect to the serialisation changes. Do we need to do something special for cluster upgrades?

felixbarny commented:
A good template for this change would be this PR:
* https://github.com/elastic/elasticsearch/pull/95558

> The main thing I'm unsure about is backwards compatibility with respect to the serialisation changes. Do we need to do something special for cluster upgrades?

Seems like that PR doesn't do anything special here with regards to the serialization changes. However, I think there may be a problematic situation that can occur in this code line in a mixed version cluster scenario (such as when doing a rolling upgrade):

https://github.com/elastic/elasticsearch/blob/b9c29807ecc831838c398aace29157d24e51fdea/server/src/main/java/org/elasticsearch/index/mapper/DynamicTemplate.java#L314

If one node in a cluster gets updated and the user sends a mapping update request to that node, it may propagate that mapping to older nodes that expect the content of `match_mapping_type` to be a string and they'd fail when an array of values is passed.

But maybe all of that is a non issue, depending on how the update process works. IIRC, at first, all non-master nodes are updated before the master node. Also, updating the mapping would go though the master node that triggers a cluster state update to propagate the mappings to other nodes. During that process, the master node would validate the mappings. I think that makes sure that ES will only accept mappings with an array in `match_mapping_type` when all nodes in the cluster are updated.

felixbarny commented on server/src/main/java/org/elasticsearch/index/mapper/DynamicTemplate.java:
```diff
@@ -257,7 +260,8 @@ static DynamicTemplate parse(String name, Map<String, Object> conf) throws Mappe
         Map<String, Object> mapping = null;
         boolean runtime = false;
-        String matchMappingType = null;
+        Object matchMappingType = null;
+        Object unmatchMappingType = null;
```
Let's use a `List<String>` here that only has one element if the `match_mapping_type`/`unmatch_mapping_type` was provided as a string.

axw replied:
Done

felixbarny commented on server/src/main/java/org/elasticsearch/index/mapper/DynamicTemplate.java:
```diff
@@ -271,7 +275,9 @@ static DynamicTemplate parse(String name, Map<String, Object> conf) throws Mappe
                 addEntriesToPatternList(pathUnmatch, propName, entry);
             } else if ("match_mapping_type".equals(propName)) {
-                matchMappingType = entry.getValue().toString();
+                matchMappingType = entry.getValue();
```
Can we use `addEntriesToPatternList` here?

axw replied:
Yes, done. Also now using `matchPatternsAreDefined` for the implicit wildcard, and made it a varargs method.

felixbarny commented on server/src/main/java/org/elasticsearch/index/mapper/DynamicTemplate.java:
```diff
@@ -553,13 +592,22 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
+                numPossibleXContentFieldTypes = XContentFieldType.values().length;
+            }
+            if (xContentFieldTypes.length < numPossibleXContentFieldTypes) {
+                builder.field("match_mapping_type", List.of(xContentFieldTypes).stream().map(XContentFieldType::toString).toList());
```
Seems like if you set `unmatch_mapping_type` and then get the mapping, it'll be modified to a `match_mapping_type`. I think we'd generally want to return the mapping as it has been provided by the user, without modifying it.

axw replied:
Agreed. I've fixed that -- I'm now passing in the user-specified `match_mapping_type` and `unmatch_mapping_type` lists, and serialising those. The final filtered list of content types is not used in the serialisation.

I have also removed the bit about skipping serialisation of `match_mapping_type` -- it will still be skipped for the implicit wildcard (because that is not touching the original lists). If a user explicitly specifies `match_mapping_type: "*"`, then that will now be serialised. Feels better/less magical to me.

axw commented on server/src/main/java/org/elasticsearch/index/mapper/DynamicTemplate.java:
```diff
@@ -301,29 +307,47 @@ static DynamicTemplate parse(String name, Map<String, Object> conf) throws Mappe
+        // unmatch_mapping_type filters down the matched mapping types.
+        if (unmatchMappingType != null) {
+            final Set<XContentFieldType> unmatchSet = Set.of(parseMatchMappingType(unmatchMappingType));
+            Set<XContentFieldType> matchSet = new LinkedHashSet<XContentFieldType>(Arrays.asList(xContentFieldTypes));
+            matchSet.removeAll(unmatchSet);
+            xContentFieldTypes = matchSet.toArray(XContentFieldType[]::new);
```
Agreed, thanks

axw committed:
Update server/src/main/java/org/elasticsearch/index/mapper/DynamicTemplate.java

Co-authored-by: Felix Barnsteiner <felixbarny@users.noreply.github.com>

Commit:
Address review comments

- Use addEntriesToPatternList like match/path_match etc.
- Serialise (un)match_mapping_type as specified, except
  for single-value lists which are always serialised as
  a non-list value. Implicit "match_mapping_type: *" is
  still not serialised, but an explicit setting will now
  be serialised where it was not before.

Commit:
Fix test

Commit:
Add YAML REST test

Commit:
Add another unit test for unmatch_mapping_type

... to highlight the original purpose, which is to
unmatch an internal object field whose path may match
a leaf field.

axw commented:
@elasticmachine run elasticsearch-ci/packaging-tests-windows-sample

axw committed:
Update docs/changelog/103171.yaml

elasticsearchmachine commented:
Hi @axw, I've created a changelog YAML for you.

felixbarny commented on server/src/main/java/org/elasticsearch/index/mapper/DynamicTemplate.java:
```diff
@@ -301,30 +307,48 @@ static DynamicTemplate parse(String name, Map<String, Object> conf) throws Mappe
-            xContentFieldTypes = new XContentFieldType[0];
+            matchXContentFieldTypes = matchMappingType.stream().map(XContentFieldType::fromString);
         }
+        if (runtime) {
+            matchXContentFieldTypes = matchXContentFieldTypes.filter(XContentFieldType::supportsRuntimeField);
+        }
```
I wonder if we should move this into the `if (wildcardMatchMappingType)` branch as the `else` branch is already guaranteed to not contain runtime-unsupported field types if `runtime` is `true`. I felt like it took me another couple of seconds to understand that when reading the code.

axw replied:
Good pick up, that's a nice simplification.

felixbarny reviewed (approved):
Just a minor suggestion, otherwise LGTM. This also needs to be reviewed by the search team before merging.

Commit:
Address review comment

Move runtime-supported type filtering.

Commit:
Update docs

elasticsearchmachine commented:
Pinging @elastic/es-search (Team:Search)

Commit:
Fix docs

axw commented:
@salvatore-campagna would you be able to review this PR? Or otherwise tag someone else more appropriate? :pray:

elasticsearchmachine commented:
Pinging @elastic/es-storage-engine (Team:StorageEngine)

salvatore-campagna commented on server/src/main/java/org/elasticsearch/index/mapper/DynamicTemplate.java:
```diff
@@ -257,7 +260,8 @@ static DynamicTemplate parse(String name, Map<String, Object> conf) throws Mappe
         Map<String, Object> mapping = null;
         boolean runtime = false;
-        String matchMappingType = null;
+        List<String> matchMappingType = new ArrayList<>(4);
```
Why is the size of the array 4? Just to have a size that is lower than he default?

felixbarny replied:
This uses the same initial size of the array list as the other lists a few lines above. I think this is because we expect most of these to have one or a few values, so a custom default value 4 aims for more efficiency than the generic default value.

salvatore-campagna commented on server/src/main/java/org/elasticsearch/index/mapper/DynamicTemplate.java:
```diff
@@ -553,13 +594,19 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
+                builder.field("unmatch_mapping_type", unmatchMappingType.get(0));
+            } else {
+                builder.field("unmatch_mapping_type", unmatchMappingType);
+            }
```
Logic seems similar...maybe extract a method just passing a list and a string?

felixbarny replied:
let me add that

salvatore-campagna commented:
LGTM, I left a question and a comment.

felixbarny committed:
Merge remote-tracking branch 'origin/main' into dynamic-template-unmatch-mapping-type

felixbarny committed:
Extract method to serialize a string or array field, depending on the size of the list
